### PR TITLE
Socint 259 flexible sample storage

### PR DESCRIPTION
**Pull request by PeterBochel**
See CR for details: https://collaborate2.ons.gov.uk/jira/browse/SOCINT-259

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java [hunk 1]
```diff
@@ -33,19 +33,30 @@
...
 
   /**
-   * the GET end point to return the latest valid Non HI Case by UPRN
+   * the GET end point to return latest valid Case which matches the supplied sample attribute
+   * name/value.
    *
-   * @param uprn the UPRN
+   * @param attributeKey - is the name of the field in the sample data to search by.
+   * @param attributeValue - is the value that target case(s) must contain.
    * @return Returned latest Non HI case with valid address for the UPRN
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java:41]: this javadoc return value description needs updating

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import uk.gov.ons.ctp.common.cloud.RetryableCloudDataStore;
+import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.common.event.model.CaseUpdate;
+
+/** A Repository implementation for CRUD operations on Case data entities */
+@Service
+public class RespondentCaseRepository {
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java:16]: All these repository names are a bit wordy. Shame you didn't just call this `CaseRepository`  and likewise with the others.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/CaseDTO.java [hunk 3]
```diff
@@ -13,5 +14,6 @@
   private boolean invalid;
   private String refusalReceived;
   private String caseRef;
-  private AddressDTO address;
+  private Map<String, String> sample;
+  private Map<String, String> sampleSensitive;
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/CaseDTO.java:18]: We've introduced the phone number here (via sampleSensitive).  This is something new. Did we want to do this ?

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java [hunk 4]
```diff
@@ -65,46 +61,41 @@ public void setUp() {
...
+        .andExpect(jsonPath("$[1].sample.townName", is(rmCase1.getSample().get("townName"))))
+        .andExpect(jsonPath("$[1].sample.postcode", is(rmCase1.getSample().get("postcode"))));
   }
 
   /** Test returns resource not found for non-existent UPRN */
   @Test
   public void getCaseByUPRNNotFound() throws Exception {
 
-    when(caseService.getLatestValidCaseByUPRN(new UniquePropertyReferenceNumber(UPRN)))
+    when(caseService.readCaseUpdateBySampleAttribute("uprn", UPRN))
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java:89]: this doesn't make any sense. This is mocking a result that would not happen. If nothing is found then an empty list is returned.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java [hunk 5]
```diff
@@ -65,46 +61,41 @@ public void setUp() {
...
         .thenThrow(new CTPException(CTPException.Fault.RESOURCE_NOT_FOUND, ERROR_MESSAGE));
 
     mockMvc
-        .perform(get("/cases/uprn/{uprn}", UPRN))
+        .perform(get("/cases/attribute/uprn/{uprn}", UPRN))
         .andExpect(status().isNotFound())
         .andExpect(jsonPath("$.error.code", is(CTPException.Fault.RESOURCE_NOT_FOUND.toString())))
         .andExpect(jsonPath("$.error.message", is(ERROR_MESSAGE)));
   }
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java:98]: consider adding another test where a different attribute other than "uprn" is used.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java [hunk 6]
```diff
@@ -33,19 +33,30 @@
...
    */
-  @RequestMapping(value = "/uprn/{uprn}", method = RequestMethod.GET)
-  public ResponseEntity<CaseDTO> getCaseByUPRN(
-      @PathVariable(value = "uprn") final UniquePropertyReferenceNumber uprn) throws CTPException {
-    log.info("Entering GET getLatestValidNonHICaseByUPRN", kv("pathParam.uprn", uprn));
-
-    CaseDTO result = caseService.getLatestValidCaseByUPRN(uprn);
-    log.debug("Exit GET getLatestValidNonHICaseByUPRN", kv("pathParam.uprn", uprn));
+  @RequestMapping(value = "attribute/{attributeKey}/{attributeValue}", method = RequestMethod.GET)
+  public ResponseEntity<List<CaseDTO>> getCaseByAttribute(
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java:45]: the method name suggests just getting one case, but we are returning a list of cases.  Suggest `listCasesByAttribute` or `findCasesByAttribute`

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java [hunk 7]
```diff
@@ -14,13 +14,16 @@
...
-   * @param uprn of address for which latest valid case details are requested
-   * @return Case details for address UPRN
+   * @param attributeKey is the name of the attribute to search on.
+   * @param attributeValue is the value that the search attribute must equal for a case to be
+   *     returned.
+   * @return a List containing matching cases. If no cases then the list will be empty.
    * @throws CTPException if anything went wrong.
    */
-  CaseDTO getLatestValidCaseByUPRN(final UniquePropertyReferenceNumber uprn) throws CTPException;
+  List<CaseDTO> readCaseUpdateBySampleAttribute(
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java:25]: This method name suggests just returning one case. Consider prefix of `listCases` or `findCases` or something

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  private final String[] searchByUprnPath = new String[] {"sample", "uprn"};
+
+  /** Setup tests */
+  @BeforeEach
+  public void setUp() throws Exception {
+    this.collectionCase = FixtureHelper.loadPackageFixtures(CaseUpdate[].class);
+    ReflectionTestUtils.setField(caseRepo, "caseSchema", "SCHEMA");
+  }
+
+  /** Returns Empty List where no valid Address cases are returned from repository */
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryTest.java:39]: consider adding unit tests for getting attributes other than UPRN

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryIT.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+public class RespondentCaseRepositoryIT extends FirestoreTestBase {
+  private static final String CASE_ID = "dc4477d1-dd3f-4c69-b181-7ff725dc9fa4";
+
+  @Autowired private RespondentCaseRepository caseRepo;
+
+  @BeforeEach
+  public void setup() throws Exception {
+    deleteAllCollections();
+  }
+
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryIT.java:23]: this would be a great place to add a real firestore test for the `readCaseUpdateBySampleAttribute` method.

**Review by rpcatling — CHANGES_REQUESTED** [review]
a few comments

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java (hunk 2 above)
Comment (philwhiles) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java:16]: I agree

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/CaseDTO.java (hunk 3 above)
Comment (philwhiles) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/representation/CaseDTO.java:18]: No harm done. UI probably won't need to look at it, and can ignore.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java (hunk 1 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java:41]: Updated.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java (hunk 4 above)
Comment (PeterBochel) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java:89]: Test fixed to match actual behaviour.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 10]
```diff
@@ -167,7 +162,7 @@ private void recordRateLimiting(
       for (Product product : products) {
         log.debug("Recording rate-limiting", kv("fulfilmentCode", product.getFulfilmentCode()));
         UniquePropertyReferenceNumber uprn =
-            UniquePropertyReferenceNumber.create(caseDetails.getSample().getUprn());
+            UniquePropertyReferenceNumber.create(caseDetails.getSample().get("uprn"));
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:165]: We should have a centralised static final def of the sample attribute names

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryTest.java (hunk 8 above)
Comment (PeterBochel) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryTest.java:39]: One of the test cases has been modified to search by postcode.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java (hunk 2 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepository.java:16]: Have renamed. Also associated test cases.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java (hunk 5 above)
Comment (PeterBochel) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpointUnitTest.java:98]: Have modified one of the tests to use a different attribute.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java (hunk 6 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java:45]: Renamed to findCasesByAttribute()

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java (hunk 7 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java:25]: Have renamed to findCasesBySampleAttribute()

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryIT.java (hunk 9 above)
Comment (PeterBochel) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/repository/RespondentCaseRepositoryIT.java:23]: New test added.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java (hunk 10 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java]: Constants for all know attribute names added to CaseUpdate in common.

**Review by rpcatling — APPROVED** [review]

**Review by philwhiles — APPROVED** [review]
